src/envs/pybullet_blocks.py

```diff
@@ -0,0 +1,754 @@
+
+        # Get block states.
+        for block_id, block in self._block_id_to_block.items():
+            (bx, by, bz), _ = p.getBasePositionAndOrientation(
+                block_id, physicsClientId=self._physics_client_id)
+            held = (block_id == self._held_block_id)
+            # pose_x, pose_y, pose_z, held
+            state_dict[block] = np.array([bx, by, bz, held], dtype=np.float32)
+
+        state = State(state_dict)
+        assert set(state) == set(self._current_state)
+
+        return state
+
+    def _detect_held_block(self) -> Optional[int]:
+        """Return the PyBullet object ID of the held object if one exists.
+
+        If multiple blocks are within the grasp tolerance, return the
+        one that is closest.
+        """
+        expected_finger_normals = {
+            self._left_finger_id: np.array([0., 1., 0.]),
+            self._right_finger_id: np.array([0., -1., 0.]),
+        }
+        closest_held_block = None
+        closest_held_block_dist = float("inf")
+        for block_id in self._block_id_to_block:
+            for finger_id, expected_normal in expected_finger_normals.items():
+                # Find points on the block that are within grasp_tol distance
+                # of the finger. Note that we use getClosestPoints instead of
+                # getContactPoints because we still want to consider the block
+                # held even if there is a tiny distance between the fingers and
+                # the block.
+                closest_points = p.getClosestPoints(
+                    bodyA=self._fetch_id,
+                    bodyB=block_id,
+                    distance=self._grasp_tol,
+                    linkIndexA=finger_id,
+                    physicsClientId=self._physics_client_id)
+                for point in closest_points:
+                    # If the contact normal is substantially different from
+                    # the expected contact normal, this is probably a block
+                    # on the outside of the fingers, rather than the inside.
+                    # A perfect score here is 1.0 (normals are unit vectors).
+                    contact_normal = point[7]
+                    if expected_normal.dot(contact_normal) < 0.5:
+                        continue
+                    # Handle the case where multiple blocks pass this check
+                    # by taking the closest one. This should be rare, but it
+                    # can happen when two blocks are stacked and the robot is
+                    # unstacking the top one.
+                    contact_distance = point[8]
+                    if contact_distance < closest_held_block_dist:
+                        closest_held_block = block_id
+                        closest_held_block_dist = contact_distance
+        return closest_held_block
+
+    ########################## Parameterized Options ##########################
+
+    def _move_end_effector_to_pose(
```

Comment:
For these four methods:
`_move_end_effector_to_pose`
`_move_end_effector_relative_to_block`
`_move_end_effector_relative_totable`
`_change_fingers`
let's rename them to `_create_*_option`, e.g.:
`_create_move_end_effector_to_pose_option`
`_create_change_fingers_option`